File: golang/pkg/agentbay/adb/adb.go
package adb

import (
	"encoding/json"
	"fmt"

	mcp "github.com/agentbay/agentbay-sdk/golang/api/client"
	"github.com/alibabacloud-go/tea/tea"
)

// Adb handles ADB operations in the AgentBay mobile environment.
type Adb struct {
	Session interface {
		GetAPIKey() string
		GetClient() *mcp.Client
		GetSessionId() string
	}
}

// NewAdb creates a new Adb object.
func NewAdb(session interface {
	GetAPIKey() string
	GetClient() *mcp.Client
	GetSessionId() string
}) *Adb {
	return &Adb{
		Session: session,
	}
}

// Shell executes an ADB shell command in the mobile environment.
func (a *Adb) Shell(command string) (string, error) {
	args := map[string]string{
		"command": command,
	}
	argsJSON, err := json.Marshal(args)
	if err != nil {
		return "", fmt.Errorf("failed to marshal args: %w", err)
	}

	callToolRequest := &mcp.CallMcpToolRequest{
		Authorization: tea.String("Bearer " + a.Session.GetAPIKey()),
		SessionId:     tea.String(a.Session.GetSessionId()),
		Name:          tea.String("shell"),
		Args:          tea.String(string(argsJSON)),
	}
	response, err := a.Session.GetClient().CallMcpTool(callToolRequest)
	if err != nil {
		return "", fmt.Errorf("failed to execute adb shell command: %w", err)
	}

	// Convert interface{} to map
	data, ok := response.Body.Data.(map[string]interface{})
	if !ok {
		return "", fmt.Errorf("invalid response data format")
	}

	// Get content field and parse as array
	contentArray, ok := data["content"].([]interface{})
	if !ok {
		return "", fmt.Errorf("content field not found or not an array")
	}

	var fullText string
	for _, item := range contentArray {
		// Assert each element is map[string]interface{}
		contentItem, ok := item.(map[string]interface{})
		if !ok {
			continue
		}

		// Extract text field
		text, ok := contentItem["text"].(string)
		if !ok {
			continue
		}

		fullText += text + "\n" // Concatenate text content
	}
	return fullText, nil
}

File: python/agentbay/adb/adb.py
import json
import traceback
from typing import Any, Dict

from agentbay.api.models import CallMcpToolRequest
from agentbay.exceptions import AdbError


class Adb:
    """
    Adb handles ADB operations in the AgentBay mobile environment.
    """

    def __init__(self, session):
        """
        Initialize the Adb object.

        Args:
            session: The session object that provides access to the AgentBay API.
        """
        self.session = session

    def shell(self, command: str) -> str:
        """
        Execute an ADB shell command in the mobile environment.

        Args:
            command: The ADB shell command to execute.

        Returns:
            The output of the ADB shell command.

        Raises:
            Exception: If the command execution fails.
        """
        args = {"command": command}

        request = CallMcpToolRequest(
            args=json.dumps(args, ensure_ascii=False),
            authorization="Bearer " + str(self.session.get_api_key()),
            name="shell",
            session_id=str(self.session.get_session_id()),
        )

        try:
            response = self.session.get_client().call_mcp_tool(request)
            respoonse_map = response.to_map()

            body_map = respoonse_map["body"]
            data_map = body_map["Data"]
            if data_map.get("isError", False):
                error_msg = json.dumps(data_map, ensure_ascii=False)
                raise AdbError(error_msg)
            content_array = data_map.get("content", [])
            text_array = []

            for item in content_array:
                try:
                    text = item.get("text", "")
                    text_array.append(str(text))
                except Exception:
                    continue
            result = "\n".join(text_array)
            return result
        except Exception as error:
            error_detail = traceback.format_exc()
            raise AdbError(
                f"Failed to execute ADB shell command: {error}\n{error_detail}"
            )
